cgen: cannot convert 'struct _option_int' to 'int'

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: linux, Ubuntu 22.04.1 LTS
Processor: 8 cpus, 64bit, little endian, 11th Gen Intel(R) Core(TM) i7-11390H @ 3.40GHz
CC version: cc (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0

getwd: /home/hitalo/Documents/bugs no v
vmodules: /home/hitalo/.vmodules
vroot: /home/hitalo/v
vexe: /home/hitalo/v/v
vexe mtime: 2022-11-17 21:00:38
is vroot writable: true
is vmodules writable: true
V full version: V 0.3.2 523ccbc.886abcd

Git version: git version 2.34.1
Git vroot status: weekly.2022.46-16-g886abcdf
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 12f392c3

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg optional_field.v`
```v
struct Test {
    a ?int = none
    b ?int
}

fn fuba<U>(val U)  {
	$for field in U.fields {
		println(val.$(field.name)) // <<< work
        variable := val.$(field.name) // <<< not work
	}
}

fn main() {
    test := Test{  }
    fuba(test)
}

```

**What did you expect to see?**

code compile


**What did you see instead?**
```
optional_field.v:9:9: warning: unused variable: `variable`
    7 |     $for field in U.fields {
    8 |         println(val.$(field.name)) // <<< work
    9 |         variable := val.$(field.name) // <<< not work
      |         ~~~~~~~~
   10 |     }
   11 | }
==================
/tmp/v_1000/optional_field.12113549329357735642.tmp.c:12126: error: cannot convert 'struct _option_int' to 'int'
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/v/gen/c/assign.v
+++ b/vlib/v/gen/c/assign.v
@@ -434,8 +434,9 @@ fn (mut g Gen) assign_stmt(node_ ast.AssignStmt) {
 				*/
 			}
 			if !cloned {
-				if (var_type.has_flag(.optional) && !val_type.has_flag(.optional))
-					|| (var_type.has_flag(.result) && !val_type.has_flag(.result)) {
+				if !g.inside_comptime_for_field
+					&& ((var_type.has_flag(.optional) && !val_type.has_flag(.optional))
+					|| (var_type.has_flag(.result) && !val_type.has_flag(.result))) {
 					tmp_var := g.new_tmp_var()
 					g.expr_with_tmp_var(val, val_type, var_type, tmp_var)
 				} else if is_fixed_array_var {

--- a/vlib/v/gen/c/comptime.v
+++ b/vlib/v/gen/c/comptime.v
@@ -121,6 +121,10 @@ fn (mut g Gen) comptime_call(mut node ast.ComptimeCall) {
 			}
 			return
 		}
+
+		if !g.inside_call && (m.return_type.has_flag(.optional) || m.return_type.has_flag(.result)) {
+			g.write('(*(${g.base_type(m.return_type)}*)')
+		}
 		// TODO: check argument types
 		g.write('${util.no_dots(sym.name)}_${g.comptime_for_method}(')
 
@@ -165,6 +169,9 @@ fn (mut g Gen) comptime_call(mut node ast.ComptimeCall) {
 			}
 		}
 		g.write(')')
+		if !g.inside_call && (m.return_type.has_flag(.optional) || m.return_type.has_flag(.result)) {
+			g.write('.data)')
+		}
 		return
 	}
 	mut j := 0

--- a/vlib/v/gen/c/testdata/comptime_optional_call.out
+++ b/vlib/v/gen/c/testdata/comptime_optional_call.out
@@ -1,0 +1,10 @@
+0
+0
+Option(0)
+Option(0)
+Option(error: none)
+Option(error: none)
+1
+1
+println(NIL)
+Option(error: none)

--- a/vlib/v/gen/c/testdata/comptime_optional_call.vv
+++ b/vlib/v/gen/c/testdata/comptime_optional_call.vv
@@ -1,0 +1,28 @@
+struct Foo {
+	foobar int
+	bar    ?int
+	baz    ?int = none
+}
+
+fn (f Foo) bar() int {
+	return 1
+}
+
+fn (f Foo) baz() ?string {
+	return none
+}
+
+fn main() {
+	foo := Foo{}
+
+	$for field in Foo.fields {
+		var := foo.$(field.name)
+		println(var)
+		println(foo.$(field.name))
+	}
+	$for method in Foo.methods {
+		var := foo.$method()
+		println(var)
+		println(foo.$method())
+	}
+}

